lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -1521,38 +1510,53 @@ bool swift::tryDeleteDeadClosure(SingleValueInstruction *closure,
     return true;
   }
 
-  // We only accept a user if it is an ARC object that can be removed if the
-  // object is dead. This should be expanded in the future. This also ensures
-  // that we are locally identified and non-escaping since we only allow for
-  // specific ARC users.
-  ReleaseTracker tracker(useDoesNotKeepClosureAlive, useHasTransitiveOwnership);
-
-  // Find the ARC users and the final retain, release.
-  if (!getFinalReleasesForValue(SILValue(closure), tracker))
+  // Collect all destroys of the closure (transitively including destorys of
+  // copies) and check if those are the only uses of the closure.
+  SmallVector<SILInstruction *, 16> closureDestroys;
+  if (!collectDestroys(closure, closureDestroys))
     return false;
 
   // If we have a partial_apply, release each captured argument at each one of
   // the final release locations of the partial apply.
   if (auto *pai = dyn_cast<PartialApplyInst>(closure)) {
-    // If we can not decrement the ref counts of the dead partial apply for any
-    // reason, bail.
-    if (!releaseCapturedArgsOfDeadPartialApply(pai, tracker, callbacks))
-      return false;
+    assert(!pa->isOnStack() &&
+           "partial_apply [stack] should have been handled before");
+    SILBuilder builder(pai);
+    if (needKeepArgsAlive) {
+      if (!keepArgsOfPartialApplyAlive(pai, closureDestroys, builder))
+        return false;
+    } else {
+      // A preceeding partial_apply -> apply conversion (done in
+      // tryOptimizeApplyOfPartialApply) already ensured that the arguments are
+      // kept alive until the end of the partial_apply's lifetime.
+      SmallVector<Operand *, 8> argsToHandle;
+      getConsumedPartialApplyArgs(pai, argsToHandle,
+                                  /*includeTrivialAddrArgs*/ false);
+
+      // We can just destroy the arguments at the point of the partial_apply
+      // (remember: partial_apply consumes all arguments).
+      for (Operand *argOp : argsToHandle) {
+        SILValue arg = argOp->get();
+        if (arg->getType().isObject()) {
+          builder.emitDestroyValueOperation(pai->getLoc(), arg);
```

Comment:
I think we do this a lot. Add this as a helper on SILBuilder? Maybe emitDestroyOperation?